lib/core/include/scipp/core/element/map_to_bins.h

```diff
@@ -0,0 +1,139 @@
+// SPDX-License-Identifier: BSD-3-Clause
+// Copyright (c) 2022 Scipp contributors (https://github.com/scipp)
+/// @file
+/// @author Simon Heybrock
+#pragma once
+#include <limits>
+
+#include "scipp/common/overloaded.h"
+#include "scipp/core/eigen.h"
+#include "scipp/core/element/arg_list.h"
+#include "scipp/core/element/util.h"
+#include "scipp/core/histogram.h"
+#include "scipp/core/subbin_sizes.h"
+#include "scipp/core/time_point.h"
+#include "scipp/core/transform_common.h"
+
+namespace scipp::core::element {
+
+auto map_to_bins_direct = [](auto &binned, auto &bins, const auto &data,
+                             const auto &bin_indices) {
+  const auto size = scipp::size(bin_indices);
+  using T = std::decay_t<decltype(data)>;
+  for (scipp::index i = 0; i < size; ++i) {
+    const auto i_bin = bin_indices[i];
+    if (i_bin < 0)
+      continue;
+    if constexpr (is_ValueAndVariance_v<T>) {
+      binned.variance[bins[i_bin]] = data.variance[i];
+      binned.value[bins[i_bin]++] = data.value[i];
+    } else {
+      binned[bins[i_bin]++] = data[i];
+    }
+  }
+};
+
+constexpr bool is_powerof2(int v) { return v && ((v & (v - 1)) == 0); }
+
+template <int chunksize>
+auto map_to_bins_chunkwise = [](auto &binned, auto &bins, const auto &data,
+                                const auto &bin_indices) {
+  // compiler is smart for div or mod 2**N, otherwise this would be too slow
+  static_assert(is_powerof2(chunksize));
+  using InnerIndex = int16_t;
+  static_assert(chunksize <= std::numeric_limits<InnerIndex>::max());
+  const auto size = scipp::size(bin_indices);
+  using T = std::decay_t<decltype(data)>;
+
+  using Val =
+      std::conditional_t<is_ValueAndVariance_v<T>, typename T::value_type, T>;
+  // TODO Ideally this buffer would be reused (on a per-thread basis)
+  // for every application of the kernel.
+  std::vector<std::tuple<std::vector<typename Val::value_type>,
+                         std::vector<InnerIndex>>>
+      chunks((bins.size() - 1) / chunksize + 1);
```

Comment:
My initial version used `reserve`, but then I switched to having an outer loop, which also reduces memory use. Therefore `reserve` is likely irrelevant / premature (except for very special cases).

I did not try an allocator. The main application of this is for binning a table, i.e., data is not binned yet. In that case every thread may just process 1 bin (or a couple of bins), so reuse does not matter.

As operating systems' allocators are getting better and better I did not feel the need to experiment (for now), but if you have concrete ideas.... last time I looked at `std::pmr` it was still very incomplete and/or difficult to use.

Reply:
I have never used it myself. But I remember a series of videos on C++ Weekly about it. I only thought about it because the inner vectors grow to a known total size, so preallocating that much memory might help. But if what you said about reserve holds in practice, then this is irrelevant.